### PR TITLE
Wire upload

### PR DESCRIPTION
**Pull request by Atlante45**

### CODE REVIEW

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -1296,7 +1295,7 @@ void Application::initializeUi() {
     rootContext->setContextProperty("Quat", new Quat());
     rootContext->setContextProperty("Vec3", new Vec3());
     rootContext->setContextProperty("Uuid", new ScriptUUID());
-    rootContext->setContextProperty("Assets", DependencyManager::get<AssetMappingsScriptingInterface>().data());
+    rootContext->setContextProperty("Assets", new AssetMappingsScriptingInterface());
```
Comment (huffman) [inline interface/src/Application.cpp:1298]: what is the reason for this change?

File: interface/src/scripting/AssetMappingsScriptingInterface.cpp [hunk 2]
```diff
@@ -46,28 +48,63 @@ void AssetMappingsScriptingInterface::getMapping(QString path, QJSValue callback
...
+    mapping = result.toString();
+
+    // Check for override
+    if (isKnownMapping(mapping)) {
+        auto message = "The following file already exists:\n" + path + "\nDo you want to override it?";
+        auto button = offscreenUi->messageBox(OffscreenUi::ICON_QUESTION, "", message,
+                                              QMessageBox::Yes | QMessageBox::No, QMessageBox::No);
+        if (button == QMessageBox::No) {
+            return;
+        }
```
Comment (birarda) [inline interface/src/scripting/AssetMappingsScriptingInterface.cpp:99]: We'll need to change this to match the styling of the other overwrite dialog.

```
var object = desktop.messageBox({
            icon: hifi.icons.question,
            buttons: OriginalDialogs.StandardButton.Yes | OriginalDialogs.StandardButton.No,
            defaultButton: OriginalDialogs.StandardButton.Yes,
            title: "Overwrite File",
            text: path + "\n" + "This file already exists. Do you want to overwrite it?"
        });
```

File: interface/src/scripting/AssetMappingsScriptingInterface.h [hunk 3]
```diff
@@ -64,6 +64,7 @@ class AssetMappingsScriptingInterface : public QObject, public Dependency {
 
     Q_INVOKABLE void setMapping(QString path, QString hash, QJSValue callback);
     Q_INVOKABLE void getMapping(QString path, QJSValue callback);
```
Comment (huffman) [inline interface/src/scripting/AssetMappingsScriptingInterface.h]: you may want to add defaults for these, I don't think you'll be able to call from the JS side without them unless they have a default value

File: interface/src/scripting/AssetMappingsScriptingInterface.cpp [hunk 4]
```diff
@@ -46,28 +48,63 @@ void AssetMappingsScriptingInterface::getMapping(QString path, QJSValue callback
...
+
+    auto upload = DependencyManager::get<AssetClient>()->createUpload(path);
+    QObject::connect(upload, &AssetUpload::finished, this, [=](AssetUpload* upload, const QString& hash) mutable {
+        if (upload->getError() != AssetUpload::NoError) {
+            if (callback.isCallable()) {
+                QJSValueList args { uint8_t(upload->getError()) };
+                callback.call(args);
+            }
+        } else {
+            setMapping(mapping, hash, callback);
```
Comment (birarda) [inline interface/src/scripting/AssetMappingsScriptingInterface.cpp:110]: Won't this cause the upload of the file to look like it failed to the asset browser if setMapping fails? I guess that's fine though it may cause extra uploads from the client.

Comment (birarda) [inline interface/src/scripting/AssetMappingsScriptingInterface.cpp:110]: There's probably not a good short term solution to this since the client doesn't know the hashes of the assets that are already present in the asset server

File: libraries/networking/src/MappingRequest.cpp [hunk 5]
```diff
@@ -109,7 +109,7 @@ void SetMappingRequest::doStart() {
     auto validPath = isValidPath(_path);
     auto validHash = isValidHash(_hash);
     if (!validPath || !validHash) {
-        _error = validPath ? MappingRequest::InvalidPath : MappingRequest::InvalidHash;
+        _error = !validPath ? MappingRequest::InvalidPath : MappingRequest::InvalidHash;
```
Comment (birarda) [inline libraries/networking/src/MappingRequest.cpp:112]: thanks

File: interface/src/Application.cpp (hunk 1 above)
Comment (Atlante45) [inline interface/src/Application.cpp:1298]: There was no need for this since it ended up being only in the QMLEngine.
I made the change be cause I thought I had to pass in the JSEngine on construction so we could call the callback with their context. Turns out you can't get that context with JSEngine.

File: interface/src/scripting/AssetMappingsScriptingInterface.cpp (hunk 4 above)
Comment (Atlante45) [inline interface/src/scripting/AssetMappingsScriptingInterface.cpp:110]: Yeah, that kinda sucks, that's why I asked you if we had upload with mapping.
I think we'll eventually want that.

File: interface/resources/qml/AssetServer.qml [hunk 6]
```diff
@@ -212,28 +212,31 @@ Window {
         });
     }
 
+    property var uploadOpen: false;
     function uploadClicked() {
+        if(uploadOpen) {
```
Comment (birarda) [inline interface/resources/qml/AssetServer.qml]: space after if

File: interface/src/scripting/AssetMappingsScriptingInterface.cpp [hunk 7]
```diff
@@ -67,28 +69,64 @@ void AssetMappingsScriptingInterface::getMapping(QString path, QJSValue callback
...
+    if (!result.isValid()) {
+        return;
+    }
+    mapping = result.toString();
+
+    // Check for override
+    if (isKnownMapping(mapping)) {
+        auto message = path + "\n" + "This file already exists. Do you want to overwrite it?";
+        auto button = offscreenUi->messageBox(OffscreenUi::ICON_QUESTION, "Overwrite File", message,
+                                              QMessageBox::Yes | QMessageBox::No, QMessageBox::No);
```
Comment (birarda) [inline interface/src/scripting/AssetMappingsScriptingInterface.cpp:96]: default the selection to yes

Comment (Atlante45) [inline interface/src/scripting/AssetMappingsScriptingInterface.cpp:96]: I don't think we want the default option to be destructive.

File: interface/src/Application.cpp (hunk 1 above)
Comment (huffman) [inline interface/src/Application.cpp:1298]: cool, I was just curious
